Add attestation evidence format support doc. (#3089)
openenclave/openenclave: docs/DesignDocs/AttestationEvidenceFormatSupport.md

=====

One of the goal of OE SDK attestation design is to be TEE-agnostic. At the same time, OE SDK should enable a verifier application to verify evidences from different sources and TEEs, if the format is known to the verifier.

Motivation
----------

Current evidence verify API `oe_verify_evidence()` verifies the evidences generated by `oe_get_evidence()` and returns the claims contained in the evidence. Such evidence has a specific format including an `oe_attestation_header`.
As reported in issue #2999, attestation service has to verify evidences from different sources other than those generated by `oe_get_evidence()`. Such as evidence generated by legacy OE SDK API `oe_get_report()` and raw TEE evidence generation such as SGX quote generation mechanism. And like `oe_verify_evidence()` returns claims, attestation service is also required to return the supplemental information. To achieve that, a new verify API is needed to verify different evidence structures and return the required supplemental information.

Goals:
- Verifier can call OE SDK API to verify known evidence formats generated by either OE SDK API or other attesters.

User Experience
---------------

> New parameter `format_id` will be added to `oe_verify_evidence` to verify various evidences other than the ones generated by `oe_get_evidence()`, such as a raw SGX quote. To verify an evidence generated by `oe_get_evidence()`, `format_id` should be `null`.

Reviewer: Is there a specific reason why the value is `null` for evidence created by `oe_get_evidence()`? It would seem more regular to assign that with its own (named) format_id, and to require format_id every time.

Reviewer 2: I think that when the evidence data is from `oe_get_evidence()`, the `format_id` parameter is optional: can be NULL or a valid format ID (which must match the one in the evidence data).

Author: > Is there a specific reason why the value is `null` for evidence created by `oe_get_evidence()`? It would seem more regular to assign that with its own (named) format_id, and to require format_id every time.

Because when verify API reads a `null` format_id, then it knows the evidence buffer is an evidence generated by `oe_get_evidence()` and will unpack the evidence first (remove oe_attestation header and extract the `format_id` in the header). Otherwise, when verify API reads a format_id, there is no solid way to tell whether the evidence buffer is a raw evidence, or an evidence generated by `oe_get_evidence`.

Reviewer: @qiucwang thank you, that makes complete sense, I'd misunderstood what happened in that case.

Scenario 1 - Verifier verifies a raw evidence (not generated by `oe_get_evidence()`) of known format id:

1. Call OE SDK public API to initialize verifier plugins.
2. Get the raw evidence and its format id from an attester, such as get a raw sgx quote from Intel SGX SDK.
3. Call the new API with the evidence and the format id to verify the evidence.

Scenario 2 - Verifier verifies an evidence generated by `oe_get_evidence()`:

1. Call OE SDK public API to initialize verifier plugins.
2. Call the new API to verify the evidence. The format id is not required because it is contained in the evidence buffer.

Scenario 1 is for users who provide their own evidence structures (like #2999, user hopes to verify a raw sgx quote).

> Scenario 2 is for users who full rely in OESDK ecosystem. They can use oe_generate_evidence to generate an oe_evidence structure and verify it by oe_verify_evidence.

Reviewer 3: ```suggestion
Scenario 2 is for users who use the existing OESDK everywhere. They can use oe_generate_evidence to generate an oe_evidence structure and verify it by oe_verify_evidence.
```
(Rationale: The OESDK "ecosystem" may include multiple formats, and in particular I hope it will include standards in the future, where the current format returned by oe_generate_evidence runs counter to emerging standards.  Hence softening wording in my suggestion.)

Author: Thank, I will update that.

In scenario 2, users should pass a `null` format id so that new `oe_get_evidence()` avoids guessing the provided evidence is a raw evidence or an oe_evidence, and the cases that the provided format id is not consistent with the id contained in the oe_evidence buffer.
```
Attester				        Verifier
--------- 				        ----------
1. SGX Quote 				        a. oe_verify_evidence(SGX_P256, evidence, ...)
2. oe_get_report() 			        b. oe_verify_evidence(OE_REPORT_ID, evidence, ...)
3. oe_get_evidence(SGX_P256) 	                c. oe_verify_evidence(null, evidence, ...)

Evidence 1 should be verified by calling function a.
Evidence 2 should be verified by calling function b.
Evidence 3 should be verified by calling function c, not a.
```

> In both scenarios, the plugin_verifier is located by format_id and then verifies the evidence. But the evidence in scenario 2 contains an extra claim buffer which is added by `oe_get_evidence()`. The plugin_verifier needs to do an extra size check to tell the difference. For example, `oe_verify_evidence(SGX_P256, evidence, ...)` will locate the verifier to verify the `evidence`. The verifier needs to check if the evidence size equals to the raw sgx quote size, or `if (evidence_size == ((sgx_quote) evidence).size)`. If it is, then there is no extra claim buffer. The verifier's claim extraction function should take care of that.

Author: updated.

Specification
-------------

### New API
An extra parameter `format_id` is added to current function oe_verify_evidence().

```
oe_result_t oe_verify_evidence(
    const oe_uuid_t* format_id,
    const uint8_t* evidence_buffer,
    size_t evidence_buffer_size,
    const uint8_t* endorsements_buffer,
    size_t endorsements_buffer_size,
    const oe_policy_t* policies,
    size_t policies_size,
    oe_claim_t** claims,
    size_t* claims_length);
```
This API is used to verify both evidences whose format id is provided by verifier and who is generated by `oe_get_evidence()`. When actual format_id is provided, it decouples evidence format ID and raw evidence buffer. When format_id is `null`, the API acts the same as previous `oe_get_evidence()`. This new design will provide more flexibility.

### New Format ID
According to issue #2999 reported by attestation service, it has to support evidence generated by SGX quote generation and `oe_get_report()`. SGX quote format id is already defined as `OE_FORMAT_UUID_SGX_ECDSA_P256`. A new id needs to be defined for evidence generated by `oe_get_report()`.

> 

Reviewer 2: Let's keep the existing `OE_FORMAT_UUID_SGX_ECDSA_P256` for the evidence generated by `oe_get_evidence()`, and define two new ones `OE_FORMAT_UUID_SGX_ECDSA_P256_REPORT` and `OE_FORMAT_UUID_SGX_ECDSA_P256_QUOTE`, for `oe_get_report()` output and raw SGX quote respectively.

I'm updating the design doc Attestation_API_Proposal.md and implementation to capture those changes.

Author: Will we have 3 different verifiers for these 3 evidences? And in the future, do we have to assigned two IDs for one evidence? One for raw evidence, another for the one generated by `oe_get_evidence()`.

Reviewer 2: I see the need of 3 format IDs. The unpacked `oe_evidence` is `oe_report`, which in turn is SGX quote prefixed with a header.

I've just created PR #3136 with the changes in Attestation_API_Proposal.md.

Author: Well my previous design set same format id for a raw evidence and oe_evidence. But I think your design makes more sense.
Here is a summary:
| Format ID                            | Evidence                                                   |
|--------------------------------------|------------------------------------------------------------|
| `OE_FORMAT_UUID_SGX_ECDSA_P256`        | SGX ECDSA-p256 evidence   generated by `oe_get_evidence()` |
| `OE_FORMAT_UUID_SGX_ECDSA_P256_QUOTE`  | raw SGX ECDSA quote                                        |
| `OE_FORMAT_UUID_SGX_ECDSA_P256_REPORT` | SGX ECDSA report generated by `oe_get_report()`            |

If we follow this way, in future, suppose we want to support new evidence formats, such as VBS, we will generate **a pair of ids**
| Format ID                | Evidence                                        |
|--------------------------|-------------------------------------------------|
| `OE_FORMAT_UUID_VBS`       | VBS evidence generated by   `oe_get_evidence()` |
| `OE_FORMAT_UUID_VBS_QUOTE` | raw VBS quote                                   |

Pros:
1. One evidence format strictly maps to one format id and verifier. It makes a cleaner design.
2. Claims extraction can be differentiated. For example, a SGX ECDSA-p256 evidence gets a `OE_CUSTOM_CLAIMS` claim entry, a raw SGX ECDSA quote gets a `OE_CLAIM_SGX_REPORT_DATA` claim entry. With different verifiers, we can call different extraction procedure for each evidence format.

Cons:
1. `oe_verify_evidence` select the plugin verifiers by `format_id` so we need to define two different verifiers for a raw evidence and the oe_evidence containing the raw evidence. We need to double the format id and verifiers even two verifiers do mostly the same thing.
2. Users might get confused about the pair of ids. We need to document it well.

Author: I've updated based on our discussion. Two new format ids are introduced: "OE_FORMAT_UUID_SGX_ECDSA_P256_QUOTE" and "OE_FORMAT_UUID_SGX_ECDSA_P256_REPORT".

Attestation Evidence Format Support
=====

One of the goal of OE SDK attestation design is to be TEE-agnostic. At the same time, OE SDK should enable a verifier application to verify evidences from different sources and TEEs, if the format is known to the verifier.

Motivation
----------

> Current evidence verify API `oe_verify_evidence()` verifies the evidence gerenated by `oe_get_evidence()` and returns the claims contained in the evidence. Such evidence has a specific format including an `oe_attestation_header`.

Reviewer 4: s/gerenated/generated

As reported in issue #2999, attestation service has to verify evidences from different sources other than those generated by `oe_get_evidence()`. Such as evidence generated by legacy OE SDK API `oe_get_report()` and raw TEE evidence generation such as SGX quote generation mechanism. And like `oe_verify_evidence()` returns claims, attestation service is also required to return the supplemental information. To achieve that, a new verify API is needed to verify different evidence structures and return the required supplemental information.

Goals:
- Verifier can call OE SDK API to verify known evidence formats generated by either OE SDK API or other attesters.

User Experience
---------------

New API will be provided to verify various evidences formats such as SGX quote.

Scenario - Verifier verifies an evidence of known format id:

1. Call OE SDK public API to initialize verifier plugins.
2. Get an evidence and its format id from attester.
3. Call the new API with the evidence and the format id to verify the evidence.
4. The new API returns attestation results and claim set.

Specification
-------------

### New API

> The new API is similar to current function oe_verify_evidence() except the extra `format_id` parameter.

Reviewer 3: Can we just replace oe_verify_evidence and keep the same name?

```
oe_result_t oe_verify_evidence_by_format_id(
    const oe_uuid_t* format_id,
    const uint8_t* evidence_buffer,
    size_t evidence_buffer_size,
    const uint8_t* endorsements_buffer,
    size_t endorsements_buffer_size,
    const oe_policy_t* policies,
    size_t policies_size,
    oe_claim_t** claims,
    size_t* claims_length);
```
This API is used to verify an evidence whose format id is provided by verifier. It decouples evidence format ID and raw evidence buffer. This API will provide more flexibility.

### New format ID
 According to issue #2999 reported by attestation service, it has to support evidence generated by SGX quote generation and `oe_get_report()`. SGX quote format id is already defined as `OE_FORMAT_UUID_SGX_ECDSA_P256`. A new id needs to be defined for evidence generated by `oe_get_report()`.

```
#define OE_FORMAT_UUID_OE_REPORT { _uuid_ }
```

### Additional Claims
Without custom claims, a new `oe_claim` field is introduced to retrieve `report_data` field inside sgx_quote. Attestation service needs this field as a part of their validation process.

```

> #define OE_CLAIM_REPORT_DATA "report_data"

Reviewer 3: can the field name also be changed to "sgx_report_data"? or does it have to interoperate with a field that already exists with the "report_data" name?

> #define OE_CLAIM_REPORT_DATA "report_data"

Author: Sure, `sgx_report_data` makes sense here.

```

Alternates
----------
We can also develop APIs that convert a raw evidence to and from an oe_evidence.

> ### convert_to_evidence().

Reviewer 3: My opinion: I don't think we need this API, we can abandon the oe_evidence going forward.

When verifier application receives a format id and a raw evidence (such as a sgx quote), `verifier.convert_to_evidence()` converts the raw evidence to an oe_evidence. Then current `oe_verify_evidence()` can verify the converted evidence.
#### Inputs:
- Evidence format ID
- Raw evidence buffer
- Raw evidence buffer size
#### Outputs:
- OE evidence buffer
- OE evidence buffer size

### convert_from_evidence().
The function converts an oe_evidence to a raw evidence for some verifier applications to handle evidence generated by `oe_get_evidence()`.
#### Inputs:
- OE evidence buffer
- OE evidence buffer size
#### Outputs:
- Evidence format ID
- Raw evidence buffer
- Raw evidence buffer size

The converion itself needs extra memory allocation and copy.

Authors
-------

> Qiucheng Wang (qiucwang.microsoft.com)

Reviewer 3: ```suggestion
Qiucheng Wang (qiucwang@microsoft.com)
```
